Skip expensive ambient light computation on fully metallic materials

Calinou opened the pull request:
`master` version of https://github.com/godotengine/godot/pull/59085.

This improves performance significantly in scenes that have fully metallic materials (metallic = 1.0).

Fully metallic *and* smooth materials (roughness = 0.0, metallic = 1.0) bring further performance improvements by also disabling shadow map sampling.

Cheap ambient light computations were left as-is to keep the shader code simple (and because no noticeable performance difference was measured while benchmarking).

**Testing project:** [test_material_roughness_performance.zip](https://github.com/godotengine/godot/files/8154719/test_material_roughness_performance.zip)

## Preview

**OS:** Fedora 34
**GPU:** NVIDIA GeForce GTX 1080
**Resolution:** 2560×1440

### Fully metallic, non-rough material

*Best-case scenario for this PR.*

#### Vulkan Clustered

| Before | After |
|-|-|
|  ![2022-02-28_16 21 10_before](https://user-images.githubusercontent.com/180032/156015709-87b7155f-d280-42d0-9635-5e91a3e9f979.png) | ![2022-02-28_16 29 11_no_shadow_no_skyambient](https://user-images.githubusercontent.com/180032/156015732-bc8e4751-ddd7-4af3-85a0-6120017b7c32.png) |

#### Vulkan Mobile

| Before | After |
|-|-|
| ![2022-03-03_20 10 10](https://user-images.githubusercontent.com/180032/156635652-147b0380-0764-4d8e-bd0c-b38e2781e538.png) | ![2022-03-03_20 08 47](https://user-images.githubusercontent.com/180032/156635643-a9d79c81-ec7d-4c29-875e-98121aa9ed1f.png) |

### Fully metallic, partially rough material

*Doesn't skip shadow map sampling, but skips environment ambient light sampling.*

#### Vulkan Clustered

| Before | After |
|-|-|
| ![2022-02-28_16 56 56](https://user-images.githubusercontent.com/180032/156015774-1c0a8b52-0931-4c78-bf3c-f6f9740fd0c2.png) | ![2022-02-28_16 57 37](https://user-images.githubusercontent.com/180032/156015802-37076014-e4a4-4b78-a4b9-9990922849e4.png) |

#### Vulkan Mobile

| Before | After |
|-|-|
| ![2022-03-03_20 10 46](https://user-images.githubusercontent.com/180032/156635665-a7a8850f-cc96-4c4a-9df0-60494013e0f5.png) | ![2022-03-03_20 08 06](https://user-images.githubusercontent.com/180032/156635608-2ba68d81-8184-4208-a2eb-ef9adfd4fc4b.png) |

### Non-metallic, rough material

*This PR does not affect non-metallic or rough materials.*

#### Vulkan Clustered

| Before | After |
|-|-|
| ![2022-02-28_16 48 31](https://user-images.githubusercontent.com/180032/156015819-e748ca2d-6227-4995-9b33-cee97c239732.png) | ![2022-02-28_16 51 27](https://user-images.githubusercontent.com/180032/156015835-04b8b38e-d168-4b6e-89b9-e37d81d51709.png) |

#### Vulkan Mobile

| Before | After |
|-|-|
| ![2022-03-03_20 10 28](https://user-images.githubusercontent.com/180032/156635655-c454f6ba-5808-4400-b11a-61417b1f513a.png) | ![2022-03-03_20 14 38](https://user-images.githubusercontent.com/180032/156636104-2dacd66d-1230-46c3-9c99-985c25d576a2.png) |

clayjohn commented:
In general this is a great idea! Both ambient and diffuse light ultimately become 0 for fully metallic materials. So we can actually skip all diffuse light and ambient light calculations when metallic is 1.

Three comments:
1. We can do the same for most of the diffuse light calculations
2. Ideally, instead of using dynamic branches, we could create a new define ``#FULLY_METALLIC`` which compiles out the ambient light and diffuse light sections. This will allow us to save a lot of VGPRs (which is one of our biggest bottlenecks right now). Of course, this would have to be available as a render_mode.
3. ``0.999`` may be too low of a constant to use and it may create some artifacts if users have a smooth gradient between metallic and non-metallic sections (this may be negligible in practice)

Calinou commented:
> Ideally, instead of using dynamic branches, we could create a new define #FULLY_METALLIC which compiles out the ambient light and diffuse light sections. This will allow us to save a lot of VGPRs (which is one of our biggest bottlenecks right now). Of course, this would have to be available as a render_mode.

The downside of that approach is that it would need more shader versions, and it wouldn't work for roughness/metallic maps that have fully white/black pixels. Therefore, I'm not sure if the tradeoff is worth it.

Calinou committed:
Skip expensive ambient light computation on fully metallic materials

This improves performance significantly in scenes that have
fully metallic materials (metallic = 1.0).

Fully metallic *and* smooth materials (roughness = 1.0, metallic = 1.0)
bring further performance improvements by also disabling shadow map
sampling.

Calinou commented:
Rebased and tested again on both Vulkan Clustered and Mobile backends, it works as expected.

I also found out that disabling SSAO and SSIL *sampling* on fully metallic materials improves performancce slightly, so I went along and did that in the clustered backend (Mobile lacks SSAO and SSIL). This also applies to rough metallic materials where neither SSAO nor SSIL are visible.

**Testing project:** [test_ssao_ssil_metallic.zip](https://github.com/godotengine/godot/files/9191591/test_ssao_ssil_metallic.zip)
